src/crimson/os/seastore/omap_manager/btree/string_kv_node_layout.h

```diff
@@ -0,0 +1,1777 @@
+      const std::string &val) {
+      buffer.push_back(
+       delta_leaf_t{
+         delta_leaf_t::op_t::UPDATE,
+         key,
+         val
+       });
+    }
+    void remove(std::string key) {
+      buffer.push_back(
+       delta_leaf_t{
+         delta_leaf_t::op_t::REMOVE,
+         key,
+         ""
+       });
+    }
+
+    void replay(StringKVLeafNodeLayout &node) {
+      for (auto &i: buffer) {
+        i.replay(node);
+      }
+    }
+    size_t get_bytes() const {
+      size_t size = 0;
+      for (auto &i: buffer) {
+        size += sizeof(i.op_t) + i.key.size() + i.val.size();
+      }
+      return size;
+    }
+    //copy out
+    void encode(ceph::bufferlist &bl) {
+      using ceph::encode;
+      uint32_t num = buffer.size();
+      encode(num, bl);
+      for (auto &&i: buffer) {
+        encode(i.op, bl);
+        encode(i.key, bl);
+        //bl.append((char*)&(i.key), sizeof(i.key));
+        encode(i.val, bl);
+      }
+      buffer.clear();
+    }
+    //copy in
+    void decode(const ceph::bufferlist &bl) {
+      using ceph::decode;
+      auto p = bl.cbegin();
+      uint32_t num;
+      decode (num, p);
+      while (num--) {
+        delta_leaf_t delta;
+        decode(delta.op, p);
+        decode(delta.key, p);
+        decode(delta.val, p);
+        buffer.push_back(delta);
+      }
+    }
+
+    bool operator==(const delta_leaf_buffer_t &rhs) const {
+      return buffer == rhs.buffer;
+    }
+  };
+
+  void journal_leaf_insert(
+    const_iterator _iter,
+    const std::string &key,
+    const std::string &val,
+    delta_leaf_buffer_t *recorder) {
+    auto iter = iterator(this, _iter.index);
+    if (recorder) {
+      recorder->insert(
+        key,
+        val);
+    }
+    leaf_insert(iter, key, val);
+  }
+
+  void journal_leaf_update(
+    const_iterator _iter,
+    const std::string &key,
+    const std::string &val,
+    delta_leaf_buffer_t *recorder) {
+    auto iter = iterator(this, _iter.index);
+    if (recorder) {
+      recorder->remove(iter->get_node_val());
+      recorder->insert(key, val);
+    }
+    leaf_update(iter, key, val);
+  }
+
+
+  void journal_leaf_remove(
+    const_iterator _iter,
+    delta_leaf_buffer_t *recorder) {
+    auto iter = iterator(this, _iter.index);
+    if (recorder) {
+      recorder->remove(iter->get_node_val());
+    }
+    leaf_remove(iter);
+  }
+
+  StringKVLeafNodeLayout(char *buf) :
+    buf(buf) {}
+
+  const_iterator iter_begin() const {
+    return const_iterator(
+      this,
+      0);
+  }
+
+  const_iterator iter_end() const {
+    return const_iterator(
+      this,
+      get_size());
+  }
+
+  iterator iter_begin() {
+    return iterator(
+      this,
+      0);
+  }
+
+  iterator iter_end() {
+    return iterator(
+      this,
+      get_size());
+  }
+
+  const_iterator iter_idx(uint16_t off) const {
+    return const_iterator(
+      this,
+      off);
+  }
+
+  const_iterator string_lower_bound(std::string str) const {
```

Comment:
Check leaf for each of the comments on inner.

Reply:
what do you mean here?